dhall/src/Dhall/Eval.hs

```diff
@@ -132,873 +105,1198 @@ instance Monoid (VChunks a) where
-vField t0 k = go t0 where
-  go = \case
-    VUnion m -> case Dhall.Map.lookup k m of
-      Just (Just _) -> VPrim $ \ ~u -> VInject m k (Just u)
-      Just Nothing  -> VInject m k Nothing
-      _             -> error errorMsg
-    VRecordLit m
-      | Just v <- Dhall.Map.lookup k m -> v
-      | otherwise -> error errorMsg
-    VProject t _ -> go t
-    VPrefer (VRecordLit m) r -> case Dhall.Map.lookup k m of
-      Just v -> VField (VPrefer (singletonVRecordLit v) r) k
-      Nothing -> go r
-    VPrefer l (VRecordLit m) -> case Dhall.Map.lookup k m of
-      Just v -> v
-      Nothing -> go l
-    VCombine (VRecordLit m) r -> case Dhall.Map.lookup k m of
-      Just v -> VField (VCombine (singletonVRecordLit v) r) k
-      Nothing -> go r
-    VCombine l (VRecordLit m) -> case Dhall.Map.lookup k m of
-      Just v -> VField (VCombine l (singletonVRecordLit v)) k
-      Nothing -> go l
-    t -> VField t k
-
-  singletonVRecordLit v = VRecordLit (Dhall.Map.singleton k v)
-{-# inline vField #-}
-
-vProjectByFields :: Eq a => Env a -> Val a -> Set Text -> Val a
+vField t0 k = go t0
+  where
+    go = \case
+        VUnion m -> case Map.lookup k m of
+            Just (Just _) -> VPrim $ \ ~u -> VInject m k (Just u)
+            Just Nothing  -> VInject m k Nothing
+            _             -> error errorMsg
+        VRecordLit m
+            | Just v <- Map.lookup k m -> v
+            | otherwise -> error errorMsg
+        VProject t _ -> go t
+        VPrefer (VRecordLit m) r -> case Map.lookup k m of
+            Just v -> VField (VPrefer (singletonVRecordLit v) r) k
+            Nothing -> go r
+        VPrefer l (VRecordLit m) -> case Map.lookup k m of
+            Just v -> v
+            Nothing -> go l
+        VCombine (VRecordLit m) r -> case Map.lookup k m of
+            Just v -> VField (VCombine (singletonVRecordLit v) r) k
+            Nothing -> go r
+        VCombine l (VRecordLit m) -> case Map.lookup k m of
+            Just v -> VField (VCombine l (singletonVRecordLit v)) k
+            Nothing -> go l
+        t -> VField t k
+
+    singletonVRecordLit v = VRecordLit (Map.singleton k v)
+{-# INLINE vField #-}
+
+vProjectByFields :: Eq a => Environment a -> Val a -> Set Text -> Val a
 vProjectByFields env t ks =
   if null ks then
     VRecordLit mempty
   else case t of
     VRecordLit kvs -> let
-      kvs' = Dhall.Map.restrictKeys kvs (Dhall.Set.toSet ks)
+      kvs' = Map.restrictKeys kvs (Dhall.Set.toSet ks)
       in VRecordLit kvs'
     VProject t' _ -> vProjectByFields env t' ks
     VPrefer l r@(VRecordLit kvs) -> let
       ksSet = Dhall.Set.toSet ks
-      kvs' = Dhall.Map.restrictKeys kvs ksSet
-      ks' = Dhall.Set.fromSet (Data.Set.difference ksSet (Dhall.Map.keysSet kvs'))
+      kvs' = Map.restrictKeys kvs ksSet
+      ks' = Dhall.Set.fromSet (Set.difference ksSet (Map.keysSet kvs'))
       in vPrefer env (vProjectByFields env l ks') (VRecordLit kvs')
     t -> VProject t (Left ks)
```

Comment:
I guess this is my recent addition in the old style. It would be nice to refactor this too, but I can also do that, if you prefer.